pyro/infer/mcmc/api.py

```diff
@@ -107,13 +107,15 @@ def run(self, *args, **kwargs):
 def _gen_samples(kernel, warmup_steps, num_samples, hook, *args, **kwargs):
     kernel.setup(warmup_steps, *args, **kwargs)
     params = kernel.initial_params
+    # yield structure (key, value.shape) of params
+    yield {k: v.shape for k, v in params.items()}
     for i in range(warmup_steps):
         params = kernel.sample(params)
         hook(kernel, params, 'warmup', i)
     for i in range(num_samples):
         params = kernel.sample(params)
         hook(kernel, params, 'sample', i)
-        yield params
+        yield torch.cat([params[site].reshape(-1) for site in sorted(params)]) if params else torch.tensor([])
```

Comment:
What do you think about having generic pack, unpack utilities?

```python
> structure, tensor = pack({"a": torch.ones(2, 2), "b": torch.tensor(2.)})
> tensors
torch.Tensor([1., 1., 1., 1., 2.])

> structure
{"a": torch.Size([2, 2]), "b": torch.Size([])}

> unpack(tensor, structure)
{"a": torch.ones(2, 2), "b": torch.tensor(2.)}
```

We can also do a `stacked_unpack` which would involve stacking the tensors and unpacking them so that the leading dim has size `len(tensors)`:

```
> stacked_unpack([torch.tensor([1., 1., 1., 1., 2.]), torch.tensor(0., 0., 0., 0., 4.)], structure)
{"a": torch.Tensor([[[1., 1.], [1., 1.]], [[0., 0.], [0., 0]]]), "b": torch.Tensor([[2.], [4.]])}
```

Reply:
I love these utilities. :D